### PR TITLE
feat(sink): kafka upsert sink with schema

### PR DESCRIPTION
**Pull request by wugouzi**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
Add an option `schema.enable` for kafka upsert sink to output with schema that is supported by kafka connect's `jsonconverter`.
Syntax:
```
create sink si_kafka_upsert_schema from t_kafka with (
    connector = 'kafka',
    properties.bootstrap.server = '127.0.0.1:29092',
    topic = 'test-rw-sink-upsert-schema',
    type = 'upsert',
    primary_key = 'id',
) format upsert encode json (
    schema.enable = true
);
```

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xiangjinwu** [comment]
Sorry for the huge conflict due to the refactor. Just to share some thoughts: it looks like this change
* only applies to json but not proto or avro
* could be extended from upsert to append-only as well
Given these, this change may be done in `JsonEncoder` rather than `UpsertFormatter`.

File: src/connector/src/sink/encoder/mod.rs [hunk 1]
```diff
@@ -90,6 +90,7 @@ pub enum TimestampHandlingMode {
 #[derive(Clone)]
 pub enum CustomJsonType {
     Doris(HashMap<String, (u8, u8)>),
+    Kafka(String),
```
Comment (xiangjinwu) [inline src/connector/src/sink/encoder/mod.rs:93]: This is not a customer json type. We could have `kafka_connect: Option<struct {root_struct_name: String}>` as a field of `JsonEncoder`.

Similarly, I do not think `CustomJsonType::Doris(HashMap<String, (u8, u8)>)` is the correct design. It should be 3 options `enum DateHandlingMode { Days, String }`, `decimal_paramater_table: HashMap<String, (u8, u8)>` and `preserve_fields_order: true`. But it is out of the scope here and we can keep it untouched.

File: src/connector/src/sink/formatter/mod.rs [hunk 2]
```diff
@@ -106,12 +106,31 @@ impl SinkFormatterImpl {
...
+                        TimestampHandlingMode::Milli,
+                        format!("{}.{}", db_name, sink_from_name),
+                    )
+                } else {
+                    JsonEncoder::new(
+                        schema.clone(),
+                        Some(pk_indices),
+                        TimestampHandlingMode::Milli,
+                    )
+                };
```
Comment (xiangjinwu) [inline src/connector/src/sink/formatter/mod.rs:123]: Could be builder style:
```
JsonEncoder::new(
  schema,
  Some(pk_indices),
  TimestampHandlingMode::Milli
).with_kafka_connect(KafkaConnectParams {
    root_struct_name: format!("{}.{}", db_name, sink_from_name)
})
```

For this to work we define the following instead:
```
pub fn with_kafka_connect(self, KafkaConnectParams) -> Self { ... }
```

**Comment by Rossil2012** [comment]
updated @xiangjinwu

File: e2e_test/sink/kafka/create_sink.slt [hunk 3]
```diff
@@ -97,6 +97,17 @@ create sink si_kafka_upsert from t_kafka with (
     primary_key = 'id',
 );
 
+statement ok
+create sink si_kafka_upsert_schema from t_kafka with (
+    connector = 'kafka',
+    properties.bootstrap.server = '127.0.0.1:29092',
+    topic = 'test-rw-sink-upsert-schema',
+    type = 'upsert',
```
Comment (xiangjinwu) [inline e2e_test/sink/kafka/create_sink.slt:105]: `type = 'upsert'` is deprecated in favor of `format upsert` below. We do not need both

File: src/connector/src/sink/encoder/json.rs [hunk 4]
```diff
@@ -262,6 +288,25 @@ fn datum_to_json_object(
     Ok(value)
 }
 
+fn json_converter_with_schema(
+    object: Value,
+    name: Option<String>,
+    fields: Vec<&Field>,
```
Comment (xiangjinwu) [inline src/connector/src/sink/encoder/json.rs:294]:
```suggestion
    name: String,
    fields: impl Iterator<Item = &Field>,
```

File: src/connector/src/sink/encoder/json.rs [hunk 5]
```diff
@@ -162,7 +187,7 @@ fn datum_to_json_object(
                 }
                 json!(v_string)
             }
-            CustomJsonType::None => {
+            _ => {
```
Comment (xiangjinwu) [inline src/connector/src/sink/encoder/json.rs:190]: revert unnecessary changes

File: src/connector/src/sink/formatter/mod.rs [hunk 6]
```diff
@@ -106,12 +107,21 @@ impl SinkFormatterImpl {
...
-                let key_encoder = JsonEncoder::new(
+                let mut key_encoder = JsonEncoder::new(
                     schema.clone(),
                     Some(pk_indices),
                     TimestampHandlingMode::Milli,
                 );
-                let val_encoder = JsonEncoder::new(schema, None, TimestampHandlingMode::Milli);
+                let mut val_encoder = JsonEncoder::new(schema, None, TimestampHandlingMode::Milli);
+
+                if matches!(format_desc.options.get("schema.enable"), Some(s) if s.to_lowercase() == "true")
```
Comment (xiangjinwu) [inline src/connector/src/sink/formatter/mod.rs:117]: It is called `schemas.enable` (with `s`) in kafka connect and can be controlled individually on key / value. @tabVersion Could you confirm if we want this option to control both, and without `s`?

File: src/connector/src/sink/encoder/json.rs [hunk 7]
```diff
@@ -262,6 +284,25 @@ fn datum_to_json_object(
     Ok(value)
 }
 
+fn json_converter_with_schema<'a>(
+    object: Value,
+    name: Option<String>,
```
Comment (xiangjinwu) [inline src/connector/src/sink/encoder/json.rs:289]: Why do we need an Option here?

**Review by xiangjinwu — COMMENTED** [review]
Please also fix the test

File: src/connector/src/sink/encoder/json.rs (hunk 7 above)
Comment (Rossil2012) [reply src/connector/src/sink/encoder/json.rs:289]: Just a copy of old code, will remove it.

File: src/connector/src/sink/encoder/mod.rs [hunk 8]
```diff
@@ -131,3 +131,8 @@ impl From<FieldEncodeError> for super::SinkError {
         Self::Encode(value.to_string())
     }
 }
+
+#[derive(Clone)]
+pub struct KafkaConnectParams {
+    pub schema_name: String,
+}
```
Comment (tabVersion) [inline src/connector/src/sink/encoder/mod.rs:138]: if you do not want to edit the member, you may use `KafkaConectParamsRef = Arc<KafkaConnectParams>`

File: src/connector/src/sink/encoder/json.rs [hunk 9]
```diff
@@ -18,21 +18,24 @@ use base64::engine::general_purpose;
...
+use super::{CustomJsonType, KafkaConnectParams, Result, RowEncoder, SerTo, TimestampHandlingMode};
+use crate::sink::formatter::debezium_json::field_to_json;
 use crate::sink::SinkError;
 
 pub struct JsonEncoder {
     schema: Schema,
     col_indices: Option<Vec<usize>>,
     timestamp_handling_mode: TimestampHandlingMode,
     custom_json_type: CustomJsonType,
+    kafka_connect: Option<KafkaConnectParams>,
```
Comment (tabVersion) [inline src/connector/src/sink/encoder/json.rs:38]: how about name it `EncodeExtParams`, not limit to kafka connect, and make it a non-option field? And we can set different defaults to each format.

File: src/connector/src/sink/formatter/mod.rs [hunk 10]
```diff
@@ -106,12 +107,31 @@ impl SinkFormatterImpl {
...
+                let mut key_encoder = JsonEncoder::new(
                     schema.clone(),
                     Some(pk_indices),
                     TimestampHandlingMode::Milli,
                 );
-                let val_encoder = JsonEncoder::new(schema, None, TimestampHandlingMode::Milli);
+                let mut val_encoder = JsonEncoder::new(schema, None, TimestampHandlingMode::Milli);
+
+                if let Some(s) = format_desc.options.get("schemas.enable") {
+                    match s.to_lowercase().as_str() {
```
Comment (tabVersion) [inline src/connector/src/sink/formatter/mod.rs:118]: parse to bool before match

File: src/connector/src/sink/encoder/json.rs (hunk 9 above)
Comment (xiangjinwu) [reply src/connector/src/sink/encoder/json.rs:38]: I am viewing this parameter as dedicated to kafka connect. If there was `EncodeExtParams`, it would look like:
```
pub struct JsonEncoder {
    schema: Schema,
    col_indices: Option<Vec<usize>>,
    extra_params: EncodeExtParams,
}
pub struct EncodeExtParams {
    timestamp_handling_mode: TimestampHandlingMode,
    custom_json_type: CustomJsonType,
    kafka_connect: Option<KafkaConnectParams>,
}
```
and the extra nesting seems unnecessary.

File: src/connector/src/sink/formatter/mod.rs (hunk 10 above)
Comment (Rossil2012) [reply src/connector/src/sink/formatter/mod.rs:118]: Like this?
```rust
                if let Some(s) = format_desc.options.get("schemas.enable") {
                    match s.to_lowercase().parse::<bool>() {
                        Ok(true) => {
                            let kafka_connect = KafkaConnectParams {
                                schema_name: format!("{}.{}", db_name, sink_from_name),
                            };
                            key_encoder = key_encoder.with_kafka_connect(kafka_connect.clone());
                            val_encoder = val_encoder.with_kafka_connect(kafka_connect);
                        }
                        Ok(false) => {}
                        _ => {
                            return Err(SinkError::Config(anyhow!(
                                "schemas.enable is expected to be `true` or `false`, got {}",
                                s
                            )));
                        }
                    }
                };
```

File: src/connector/src/sink/encoder/json.rs [hunk 11]
```diff
@@ -262,6 +283,95 @@ fn datum_to_json_object(
...
+                    .types()
+                    .enumerate()
+                    .map(|(idx, sub_type)| (format!("f{}", idx + 1), sub_type))
+                    .for_each(map_sub_struct);
+            } else {
+                struct_type
+                    .iter()
+                    .map(|(sub_name, sub_type)| (sub_name.to_string(), sub_type))
+                    .for_each(map_sub_struct);
+            }
```
Comment (xiangjinwu) [inline src/connector/src/sink/encoder/json.rs:358]: Good catch. This logic of handling of struct with unnamed fields should be applicable to other places as well. For now let's just ignore them and use the same call as others: `struct_type.iter()`

Comment (Rossil2012) [reply src/connector/src/sink/encoder/json.rs:358]: Do you mean that I should delete the `if struct_type.is_unnamed()` branch, and keep only:
```rust
                struct_type
                    .iter()
                    .map(|(sub_name, sub_type)| (sub_name.to_string(), sub_type))
                    .for_each(map_sub_struct);
```

Comment (xiangjinwu) [reply src/connector/src/sink/encoder/json.rs:358]: Yes. Or in a slightly better form: `map().collect()` rather than `for_each()` on `mut Vec`

File: src/connector/src/sink/encoder/json.rs [hunk 12]
```diff
@@ -262,6 +283,84 @@ fn datum_to_json_object(
...
+    mapping.insert(
+        "schema".to_string(),
+        json!({
+            "type": "struct",
+            "fields": fields.map(|field| {
+                let mut mapping = type_as_json_schema(&field.data_type);
+                mapping.insert("field".to_string(), json!(field.name));
+                mapping
+            }).collect_vec(),
+            "optional": "false",
```
Comment (xiangjinwu) [inline src/connector/src/sink/encoder/json.rs:301]: `"optional": false,` without quotes

File: src/connector/src/sink/encoder/json.rs [hunk 13]
```diff
@@ -109,7 +130,7 @@ impl SerTo<String> for Value {
     }
 }
 
-fn datum_to_json_object(
+pub fn datum_to_json_object(
```
Comment (xiangjinwu) [inline src/connector/src/sink/encoder/json.rs:133]: unnecessary change?

File: src/connector/src/sink/encoder/json.rs [hunk 14]
```diff
@@ -262,6 +283,84 @@ fn datum_to_json_object(
...
+    }
+
+    mapping
+}
+
+// reference: https://github.com/apache/kafka/blob/80982c4ae3fe6be127b48ec09caff11ab5f87c69/connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java#L356
+fn field_as_json_schema(field: &Field) -> Value {
+    Value::Object(type_as_json_schema(&field.data_type))
+}
+
```
Comment (xiangjinwu) [inline src/connector/src/sink/encoder/json.rs:363]: no longer used?

File: src/connector/src/sink/formatter/mod.rs [hunk 15]
```diff
@@ -18,7 +18,7 @@ use risingwave_common::array::StreamChunk;
 use crate::sink::{Result, SinkError};
 
 mod append_only;
-mod debezium_json;
+pub mod debezium_json;
```
Comment (xiangjinwu) [inline src/connector/src/sink/formatter/mod.rs:21]: unnecessary change?

**Review by xiangjinwu — APPROVED** [review]
@tabVersion to check if your comments are resolved

File: src/connector/src/sink/encoder/json.rs (hunk 9 above)
Comment (tabVersion) [reply src/connector/src/sink/encoder/json.rs]: > I am viewing this parameter as dedicated to kafka connect. 

I think there will be more and more options and we cannot always add to encoder props. An extra Params struct can apply to other encoders with different default values. 

```rust
pub struct EncodeExtParams {
    timestamp_handling_mode: TimestampHandlingMode,
    custom_json_type: CustomJsonType,
    kafka_connect: Option<KafkaConnectParams>,
}

impl EncodeExtParams {
    const DEBEZIUM = EncodeExtParams { .. }
    const UPSERT = EncodeExtParams { .. }
}
```

Comment (xiangjinwu) [reply src/connector/src/sink/encoder/json.rs]: > An extra Params struct can apply to other encoders with different default values.

First of all, we are discussing about `encoder` params rather than `formatter` params. So a proper example is whether json/proto/avro/csv/template would share a lot of parameters. It would be an anti-pattern to:
```
pub struct EncodeExtParams {
    // only json and proto, not avro
    timestamp_handling_mode: TimestampHandlingMode,
    // only json
    custom_json_type: CustomJsonType,
    // json, proto, and avro, but not template
    kafka_connect: Option<KafkaConnectParams>,
    // csv
    delimiter: &str,
    // only template
    template: &str,
}
```

But it is just a difference in our predictions of future. If it turns out to be more common options, I agree they should be extracted to a common struct.

**Comment by tabVersion** [comment]
> @tabVersion to check if your comments are resolved

LGTM

**Review by tabVersion — APPROVED** [review]
